source/Components/Button.cpp

```diff
@@ -0,0 +1,328 @@
+        this->updateAligment();
+        this->updateStyle();
+        this->onClick(&Button::setFocused, this, false);
+    }
+
+    Button::~Button() {}
+
+    void Button::canBeActive(bool active, bool automatic)
+    {
+        if (active == this->isActive()) {
+            return;
+        }
+
+        if (active) {
+            this->activeSignalId = automatic ? this->onClick(&Button::toggleActive, this) : 1;
+        } else {
+            if (this->activeSignalId > 1) {
+                this->onClick.disconnect(this->activeSignalId);
+            }
+
+            this->activeSignalId = 0;
+        }
+    }
+
+    void Button::setSize(Size size)
+    {
+        this->size = size;
+        if (size == Size::Small) {
+            this->setHeight(24);
+            this->setImageSize(12);
+            this->setTextSize(10);
+        } else if (size == Size::Normal) {
+            this->setHeight(36);
+            this->setImageSize(15);
+            this->setTextSize(12);
+        } else if (size == Size::Large) {
+            this->setHeight(48);
+            this->setImageSize(18);
+            this->setTextSize(15);
+        }
+    }
+
+    void Button::setSize(const tgui::Layout2d& size)
+    {
+        if (this->getImageScaling() >= 0) {
+            tgui::BitmapButton::setSize(size);
+            return;
+        }
+
+        tgui::Button::setSize(size);
+        this->recalculateGlyphSize();
+    }
+
+    void Button::updateSize()
+    {
+        if (this->getImageScaling() >= 0) {
+            tgui::BitmapButton::updateSize();
+            return;
+        }
+
+        if (!this->imageComponent->isVisible()) {
+            tgui::Button::updateSize();
+            return;
+        }
+
+        auto size = tgui::BitmapButton::getSize();
+        if (this->m_autoSize) {
+            this->m_autoLayout = tgui::AutoLayout::Manual;
+
+            const tgui::Outline &borders = this->m_backgroundComponent->getBorders();
+
+            Widget::setSize({
+                size.x,
+                std::round(this->m_textComponent->getLineHeight() * 1.25f) + borders.getTop() + borders.getBottom()
+            });
+
+            this->recalculateGlyphSize();
+
+            if (this->m_string.empty())
+            {
+                const tgui::Vector2f innerSize = this->m_backgroundComponent->getClientSize();
+
+                Widget::setSize({
+                    this->imageComponent->getSize().x + (innerSize.y - this->imageComponent->getSize().y) + borders.getLeft() + borders.getRight(),
+                    size.y
+                });
+            } else {
+                const float spaceAroundImageAndText = this->m_textComponent->getLineHeight();
+
+                Widget::setSize({
+                    this->imageComponent->getSize().x + this->distanceBetweenTextAndImage + this->m_textComponent->getSize().x + spaceAroundImageAndText + borders.getLeft() + borders.getRight(),
+                    size.y
+                });
```

Comment:
Instead of 2 calls to `setSize`, could we compute the `size.x` and `size.y`, store them in variables and call `setSize` once at the end ?

Reply:
As mentioned above: the Button class was really complicated, so I preferred to keep the code as similar as possible to the base function code, in order to be able to modify these functions if an update to the base class code is released and is important to apply